packages/workbox-window/Workbox.mjs

```diff
@@ -0,0 +1,417 @@
+          logger.warn('An external service worker has installed. ' +
+              'You may want to suggest users reload this page.');
+        }
+        this._dispatchEvent('externalInstalled', sw);
+      } else {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.log('Registered service worker installed.');
+        }
+        this._dispatchEvent('installed', sw);
+      }
+    };
+
+    const onWaiting = () => {
+      if (isExternal) {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.warn('An external service worker has installed but is ' +
+              'waiting for this client to close before activating...');
+        }
+        this._dispatchEvent('externalWaiting', sw);
+      } else {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.warn('The service worker has installed but is waiting for ' +
+              'existing clients to close before activating...');
+        }
+        this._dispatchEvent('waiting', sw);
+      }
+    };
+
+    const onActivated = () => {
+      if (isExternal) {
+        this._dispatchEvent('externalActivated', sw);
+        if (process.env.NODE_ENV !== 'production') {
+          logger.warn('An external service worker has activated.');
+        }
+      } else {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.log('Registered service worker activated.');
+        }
+        this._dispatchEvent('activated', sw);
+      }
+    };
+
+    const onControlling = () => {
+      if (!isExternal) {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.log('Registered service worker now controlling this page.');
+        }
+        this._dispatchEvent('controlling', sw);
+      }
+    };
+
+    const onRedundant = () => {
+      if (!isExternal) {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.log('Registered service worker now redundant!');
+        }
+        this._dispatchEvent('redundant', sw);
+      }
+      sw.removeEventListener('statechange', this._onStateChange);
+    };
+
+    switch (state) {
+      case 'installed':
+        onInstalled();
+
+        // This timeout is used to ignore cases where the service worker calls
+        // `skipWaiting()` in the install event, thus moving it directly in the
+        // activating state. (Since all service workers *must* go through the
+        // waiting phase, the only way to detect `skipWaiting()` called in the
+        // install event is to observe that the time spent in the waiting phase
+        // is very short.)
+        // NOTE: we don't need separate timeouts for the new and updated SWs
+        // since they can't go through these phases at the same time.
+        this._waitingTimeout = setTimeout(() => {
+          // Ensure the SW is still waiting (it may now be redundant).
+          if (state === 'installed' && this._registration.waiting === sw) {
+            onWaiting();
+          }
+        }, 100);
```

Comment:
Can you put this `100` into a constant somewhere and document what it's used for? It's kind of magic as it is.

Reply:
Yeah, good call.